### PR TITLE
Youtube Plugin Fix

### LINKED ISSUE #2750: Youtube plugin not working with some vods
## Plugin Issue

- [x ] This is a plugin issue and I have read the contribution guidelines.

### Description

On some vods i get no playable streams found and cant seem to identify why.

### Reproduction steps / Explicit stream URLs to test

1. Try to load streamlink https://www.youtube.com/watch?v=FmDUUGOTFCM

### Log output

```
streamlink https://www.youtube.com/watch?v=jXg93FFZXWk -l debug
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.6.6
[cli][debug] Streamlink: 1.3.0
[cli][debug] Requests(2.22.0), Socks(1.7.1), Websocket(0.56.0)
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=jXg93FFZXWk
[plugin.youtube][debug] Video ID from URL
[plugin.youtube][debug] Using video ID: jXg93FFZXWk
[plugin.youtube][debug] get_video_info - 1: Found data
[plugin.youtube][debug] MuxedStream: v 137 a 251 = 1080p
Available streams: audio_mp4, audio_webm, 360p (worst), 720p, 1080p (best)

streamlink https://www.youtube.com/watch?v=FmDUUGOTFCM -l debug
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.6.6
[cli][debug] Streamlink: 1.3.0
[cli][debug] Requests(2.22.0), Socks(1.7.1), Websocket(0.56.0)
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=FmDUUGOTFCM
[plugin.youtube][debug] Video ID from URL
[plugin.youtube][debug] Using video ID: FmDUUGOTFCM
[plugin.youtube][debug] get_video_info - 1: Found data
error: No playable streams found on this URL: https://www.youtube.com/watch?v=FmDUUGOTFCM
```

### Additional comments, screenshots, etc.

[Love Streamlink? Please consider supporting our collective. Thanks!](https://opencollective.com/streamlink/donate)

### PR DESCRIPTION
**Pull request by brouxco**
Modified the configuration schema because Youtube API changed.

It works.
I kept a lot of the plugin code and got rid of the old useless code.
```adp_audio``` may not be accurate and ```adp_video``` is incomplete.

I had a look at #2822. @fgietzen are you sure ffmpeg is installed on your system ?

I'm not sure who is in charge of this module. Maybe @beardypig ?

### CODE REVIEW

**Comment by brouxco** [comment]
Side-effect: resolves #1158

**Comment by fgietzen** [comment]
> I had a look at #2822. @fgietzen are you sure ffmpeg is installed on your system ?

Unfortunately not, I have not thought of that.

File: src/streamlink/plugins/youtube.py [hunk 1]
```diff
@@ -354,44 +331,37 @@ def _get_streams(self):
...
         streams = {}
-        protected = False
-        for stream_info in info.get("url_encoded_fmt_stream_map", []):
-            if stream_info.get("s"):
-                protected = True
-                continue
+        log.debug("{0}".format(info))
+        if (info.get("player_response", {}).get("streamingData", {}).get("adaptiveFormats", [])[0].get("cipher")
+           or info.get("player_response", {}).get("streamingData", {}).get("formats", [])[0].get("cipher")):
+            log.debug("This video is protected.")
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:342]: **won't work**
`[plugin.youtube][debug] This video is protected.`

```
$ streamlink --plugin-dirs gg/ https://www.youtube.com/watch?v=24_dtEGEYCI -l info
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=24_dtEGEYCI
error: This plugin does not support protected videos, try youtube-dl instead
```

current version would work

```
$ streamlink https://www.youtube.com/watch?v=24_dtEGEYCI -l info
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=24_dtEGEYCI
[cli][info] Available streams: 144p (worst), 240p, 360p, 480p, 720p, 1080p (best)
[cli][info] Opening stream: 1080p (hls)
[cli][info] Starting player: /usr/bin/mpv
```

**Comment by beardypig** [comment]
@brouxco thanks for the contribution, code wise it looks good - I will test it out later.

File: src/streamlink/plugins/youtube.py [hunk 2]
```diff
@@ -354,44 +331,37 @@ def _get_streams(self):
...
+            raise PluginError("This plugin does not support protected videos, "
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:343]: maybe this is called to early

File: src/streamlink/plugins/youtube.py [hunk 3]
```diff
@@ -38,53 +38,32 @@ def parse_fmt_list(formatsmap):
...
+                        "mimeType": validate.text,
+                        validate.optional("url"): validate.text,
+                        validate.optional("cipher"): validate.text,
+                        validate.optional("qualityLabel"): validate.text,
+                        validate.optional("bitrate"): int
+                    }]
                 },
                 validate.optional("videoDetails"): {
-                    validate.optional("isLive"): validate.transform(bool),
+                    validate.optional("isLiveContent"): validate.transform(bool),
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:63]: are you sure this changed as it is working normally right now with `isLive`

File: src/streamlink/plugins/youtube.py (hunk 1 above)
Comment (brouxco) [reply src/streamlink/plugins/youtube.py:347]: Urls of your stream are actually ciphered (see [24_dtEGEYCI.json](https://gist.github.com/brouxco/68f02161fd7f8dc5779ac173b14f1b41))
It is not yet supported in streamlink AFAIK

Comment (brouxco) [reply src/streamlink/plugins/youtube.py]: I changed it (see 00576f98)

File: src/streamlink/plugins/youtube.py (hunk 2 above)
Comment (brouxco) [reply src/streamlink/plugins/youtube.py]: Modified (see 00576f98)

File: src/streamlink/plugins/youtube.py [hunk 4]
```diff
@@ -354,35 +336,32 @@ def _get_streams(self):
...
-            if stream_info.get("s"):
-                protected = True
-                continue
+        if (info.get("player_response", {}).get("streamingData", {}).get("adaptiveFormats", [dict()])[0].get("cipher")
+           or info.get("player_response", {}).get("streamingData", {}).get("formats", [dict()])[0].get("cipher")):
+            protected = True
+            log.debug("This video may be protected.")
 
+        for stream_info in info.get("player_response", {}).get("streamingData", {}).get("formats", []):
             stream = HTTPStream(self.session, stream_info["url"])
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:352]:
```
$ streamlink https://www.youtube.com/watch?v=Nj2U6rhnucI
[cli][debug] OS:         Linux-5.3.0-45-generic-x86_64-with-Ubuntu-19.10-eoan
[cli][debug] Python:     2.7.15+
[cli][debug] Streamlink: 1.3.1+61.g9548a69.dirty
[cli][debug] Requests(2.22.0), Socks(1.6.7), Websocket(0.54.0)
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=Nj2U6rhnucI
[plugin.youtube][debug] Video ID from URL
[plugin.youtube][debug] Using video ID: Nj2U6rhnucI
[plugin.youtube][debug] get_video_info - 1: Found data
[plugin.youtube][debug] This video may be protected.
Traceback (most recent call last):
  File "py2/bin/streamlink", line 11, in <module>
    load_entry_point('streamlink', 'console_scripts', 'streamlink')()
  File "src/streamlink_cli/main.py", line 1050, in main
    handle_url()
  File "src/streamlink_cli/main.py", line 587, in handle_url
    streams = fetch_streams(plugin)
  File "src/streamlink_cli/main.py", line 467, in fetch_streams
    sorting_excludes=args.stream_sorting_excludes)
  File "src/streamlink/plugin/plugin.py", line 317, in streams
    ostreams = self._get_streams()
  File "src/streamlink/plugins/youtube.py", line 352, in _get_streams
    stream = HTTPStream(self.session, stream_info["url"])
KeyError: u'url'
```

same for `_create_adaptive_streams`

File: src/streamlink/plugins/youtube.py [hunk 5]
```diff
@@ -354,35 +336,32 @@ def _get_streams(self):
...
-            name = formats.get(stream_info["itag"]) or stream_info["quality"]
-
-            if stream_info.get("stereo3d"):
-                name += "_3d"
+            name = stream_info["qualityLabel"]
 
             streams[name] = stream
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:355]: there are also duplicates HTTP streams
_not directly related to your PR, just a small bug in Streamlink._

File: src/streamlink/plugins/youtube.py [hunk 6]
```diff
@@ -325,16 +306,17 @@ def _get_stream_info(self, video_id):
...
+            playability_status = player_response.get("playabilityStatus")
+            if (playability_status.get("status") != "OK"):
+                reason = playability_status.get("reason")
                 log.debug("get_video_info - {0}: {1}".format(
                     count, reason)
                 )
                 continue
-            self.author = info_parsed.get("author")
-            self.title = info_parsed.get("title")
+            video_details = player_response.get("videoDetails")
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:317]: `video_details =`

not really needed here, just use it in _author_ and _title_

File: src/streamlink/plugins/youtube.py [hunk 7]
```diff
@@ -38,53 +38,40 @@ def parse_fmt_list(formatsmap):
 
 _config_schema = validate.Schema(
     {
-        validate.optional("fmt_list"): validate.all(
-            validate.text,
-            validate.transform(parse_fmt_list)
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:43]: remove `def parse_fmt_list` `def parse_stream_map`

File: src/streamlink/plugins/youtube.py (hunk 4 above)
Comment (brouxco) [reply src/streamlink/plugins/youtube.py:335]: Beginner's mistake on my part. 
I'm fixing it.
We should probably improve the unit test for the youtube plugin (in another PR). 
Here is a good resource (https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/youtube.py#L539)

**Comment by bastimeyer** [comment]
You need to rebase to master, as we've migrated the CI to Github today.

File: src/streamlink/plugins/youtube.py (hunk 5 above)
Comment (brouxco) [reply src/streamlink/plugins/youtube.py:338]: I'm not really sure what you want me to do here.
Should I check if the HTTP stream is already in the ```streams```?

**Comment by brouxco** [comment]
I did it anyway. It looks like it's ok

**Comment by bastimeyer** [comment]
The reason why I asked for a rebase was that I only saw a simple commit on top of the old PR base. I didn't notice that it was still using the new CI config for whatever reason - don't understand why tbh. Maybe because of the master branch protection and merge requirements, still weird though.

Now you've included the commits from master in the PR, which isn't good:
https://github.com/streamlink/streamlink/pull/2858/commits

You should do a proper rebase of your PR branch onto the up2date master branch and then force-push. I'm asking because I'm not sure if the current state of this PR would overwrite the CI config again if it gets squash-merged after the other CI cleanup PR that's currently still open was merged.

I don't want to cause any confusion here though, so just keep it as it is right now and fix up the stuff later once everything plugin-related is in the right shape.

File: src/streamlink/plugins/youtube.py [hunk 8]
```diff
@@ -231,24 +194,24 @@ def get_oembed(self):
...
+    def _create_adaptive_streams(self, info, streams):
         adaptive_streams = {}
         best_audio_itag = None
 
-        # Extract audio streams from the DASH format list
-        for stream_info in info.get("adaptive_fmts", []):
-            if stream_info.get("s"):
-                protected = True
+        # Extract audio streams from the adaptive format list
+        streaming_data = info.get("player_response").get("streamingData", dict())
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:202]:
```suggestion
        streaming_data = info.get("player_response", {}).get("streamingData", {})
```

File: src/streamlink/plugins/youtube.py [hunk 9]
```diff
@@ -325,16 +288,16 @@ def _get_stream_info(self, video_id):
...
+            if (playability_status.get("status") != "OK"):
+                reason = playability_status.get("reason")
                 log.debug("get_video_info - {0}: {1}".format(
                     count, reason)
                 )
                 continue
-            self.author = info_parsed.get("author")
-            self.title = info_parsed.get("title")
+            self.author = player_response.get("videoDetails").get("author")
+            self.title = player_response.get("videoDetails").get("title")
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:300]:
```suggestion -2+2
            self.author = player_response.get("videoDetails", {}).get("author")
            self.title = player_response.get("videoDetails", {}).get("title")
```

File: src/streamlink/plugins/youtube.py [hunk 10]
```diff
@@ -325,16 +288,16 @@ def _get_stream_info(self, video_id):
 
             res = self.session.http.get(self._video_info_url, params=params)
             info_parsed = parse_query(res.content if is_py2 else res.text, name="config", schema=_config_schema)
-            if (info_parsed.get("player_response", {}).get("playabilityStatus", {}).get("status") != "OK"
-                    or info_parsed.get("status") == "fail"):
-                reason = (info_parsed.get("player_response", {}).get("playabilityStatus", {}).get("reason")
-                          or info_parsed.get("reason"))
+            player_response = info_parsed.get("player_response")
+            playability_status = player_response.get("playabilityStatus")
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:292]:
```
    playability_status = player_response.get("playabilityStatus")
AttributeError: 'NoneType' object has no attribute 'get'
```

if you want to do it in two lines do it like this

```suggestion
            player_response = info_parsed.get("player_response", {})
            playability_status = player_response.get("playabilityStatus", {})
```

File: src/streamlink/plugins/youtube.py [hunk 11]
```diff
@@ -354,35 +317,34 @@ def _get_streams(self):
...
 
-        formats = info.get("fmt_list")
         streams = {}
         protected = False
-        for stream_info in info.get("url_encoded_fmt_stream_map", []):
-            if stream_info.get("s"):
-                protected = True
-                continue
+        if (info.get("player_response", {}).get("streamingData", {}).get("adaptiveFormats", [dict()])[0].get("cipher")
+           or info.get("player_response", {}).get("streamingData", {}).get("formats", [dict()])[0].get("cipher")):
```
Comment (back-to) [inline src/streamlink/plugins/youtube.py:328]: `{}` should also work for `dict()` or ?

Comment (brouxco) [reply src/streamlink/plugins/youtube.py:328]: True

**Comment by brouxco** [comment]
Do you want me to fix something else in my PR?

**Pull request merged**